web/modules/custom/asu_statistics/asu_statistics.module

```diff
@@ -0,0 +1,149 @@
+ */
+function asu_statistics_file_download($uri) {
+  if (preg_match('/asu_statistics_/', $uri)) {
+    if (\Drupal::currentUser()->hasPermission('view islandora repository reports')) {
+      return ['Content-disposition' => 'attachment; filename="' . basename($uri) . '"'];
+    }
+    else {
+      return -1;
+    }
+  }
+  return NULL;
+}
+
+/**
+ * Returns build array from data - filling in any gaps with 0s.
+ *
+ * @param array $result
+ *   The incoming data array that will define the build array.
+ */
+function asu_statistics_make_table_rows_from_result(array $result) {
+  $build_output = $rows = [];
+  $month_names = [];
+  for ($month = 1; $month < 13; $month++) {
+    $month_names[$month] = date("F", mktime(0, 0, 0, $month, 10));
+  }
+  // Get all of the rows into an array where each row contains the results for
+  // the query result which have records with these years / months.
+  $earliest_year = PHP_INT_MAX;
+  $latest_year = 0;
+  foreach ($result as $child_obj) {
+    if ($child_obj->item_year < $earliest_year) {
+      $earliest_year = $child_obj->item_year;
+    }
+    if ($child_obj->item_year > $latest_year) {
+      $latest_year = $child_obj->item_year;
+    }
+    $record_arr = (array) $child_obj;
+    $record_arr['created'] = mktime(0, 0, 0, $child_obj->item_month, 10);
+    $record_arr['item_month_name'] = $month_names[$child_obj->item_month];
+    $pad0_month = (($child_obj->item_month < 10) ? "0" : "") . $child_obj->item_month;
+    $rows[$child_obj->item_year . "-" . $pad0_month] = $record_arr;
+  }
+
+  // Fill any empty months with a 0.
+  for ($year = $earliest_year; $year <= $latest_year; $year++) {
+    $build_output[$year]['Year'] = $year;
+    for ($month = 1; $month < 13; $month++) {
+      $pad0_month = (($month < 10) ? "0" : "") . $month;
+      $year_month_key = $year . "-" . $pad0_month;
+      $month_key = substr($month_names[$month], 0, 3);
+      if (!array_key_exists($year_month_key, $rows)) {
+        $build_output[$year][$month_key] = 0;
+      }
+      else {
+        $build_output[$year][$month_key] = $rows[$year_month_key]['items'];
+      }
+    }
+  }
+  // now, add a row total at the end.
+  foreach ($build_output as $year => $month_data) {
+    $row_total = 0;
+    $skip_first_cell = TRUE;
+    foreach ($month_data as $items) {
+      $row_total += ($skip_first_cell) ? 0 : $items;
+      $skip_first_cell = FALSE;
+    }
+
+    $build_output[$year]['Total'] = $row_total;
+  }
+  return $build_output;
+}
+
+/**
+ * Returns the collection item accession stats grouped by month and year.
+ *
+ * @param mixed $collection_node_id
+ *   When provided, this integer is the collection's node id.
+ * @param bool $status
+ *   Whether or not to limit the query to published items.
+ * @param bool $count_items
+ *   If not provided, the query will return collections for the site-report.
+ */
+function asu_statistics_get_stats($collection_node_id = NULL, $status = TRUE, $count_items = TRUE) {
```

Comment:
if this function is only called by the controller - should it live in the controller?